pallets/stablewap/src/lib.rs

```diff
@@ -0,0 +1,696 @@
+		///
+		/// new_reserve_b = (reserve_a + amount) * reserve_b / reserve_a
+		/// amount_b = new_reserve_b - reserve_b
+		///
+		/// LP must have sufficient amount of asset a/b - amount_a and amount_b.
+		///
+		/// Origin is given corresponding amount of shares.
+		///
+		/// Parameters:
+		/// - `origin`: liquidity provider
+		/// - `pool_id`: Pool Id
+		/// - `asset`: Asset id
+		/// - `amount`: liquidity amount of `asset` to be added to the pool.
+		///
+		/// Emits `LiquidityAdded` event when successful.
+		#[pallet::weight(<T as Config>::WeightInfo::add_liquidity())]
+		#[transactional]
+		pub fn add_liquidity(
+			origin: OriginFor<T>,
+			pool_id: PoolId<T::AssetId>,
+			asset: T::AssetId,
+			amount: Balance,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				amount >= T::MinTradingLimit::get(),
+				Error::<T>::InsufficientTradingAmount
+			);
+
+			// Ensure that who has enough balance of given asset
+			ensure!(
+				T::Currency::free_balance(asset, &who) >= amount,
+				Error::<T>::InsufficientBalance
+			);
+
+			let pool = Pools::<T>::get(pool_id).ok_or(Error::<T>::PoolNotFound)?;
+
+			ensure!(pool.contains_asset(asset), Error::<T>::AssetNotInPool);
+
+			let pool_account = T::ShareAccountId::from_assets(&pool.assets, Some(POOL_IDENTIFIER));
+
+			// Work out which asset is asset B ( second asset which has to provided by LP )
+			// Update initial reserves in correct order ( asset a is given as params, asset b is second asset)
+			let (assets, initial_reserves) = {
+				let initial = (
+					T::Currency::free_balance(pool.assets.0, &pool_account),
+					T::Currency::free_balance(pool.assets.1, &pool_account),
+				);
+				if asset == pool.assets.0 {
+					((asset, pool.assets.1), AssetAmounts(initial.0, initial.1))
+				} else {
+					((asset, pool.assets.0), AssetAmounts(initial.1, initial.0))
+				}
+			};
+
+			let updated_a_reserve = initial_reserves
+				.0
+				.checked_add(amount)
+				.ok_or(ArithmeticError::Overflow)?;
+
+			// Calculate new reserve of asset b and work out the difference to get required amount of second asset which has to be provided by LP too.
+			let asset_b_reserve = calculate_asset_b_reserve(initial_reserves.0, initial_reserves.1, updated_a_reserve)
+				.ok_or(ArithmeticError::Overflow)?;
+
+			let asset_b_amount = asset_b_reserve
+				.checked_sub(initial_reserves.1)
+				.ok_or(ArithmeticError::Underflow)?;
+
+			ensure!(
+				T::Currency::free_balance(assets.1, &who) >= asset_b_amount,
+				Error::<T>::InsufficientBalance
+			);
+
+			let new_reserves = AssetAmounts(updated_a_reserve, asset_b_reserve);
+
+			let share_issuance = T::Currency::total_issuance(pool_id.0);
+
+			let share_amount = calculate_add_liquidity_shares(
+				&initial_reserves,
+				&new_reserves,
+				T::Precision::get(),
+				pool.amplification.into(),
+				share_issuance,
+			)
+			.ok_or(ArithmeticError::Overflow)?;
+
+			T::Currency::deposit(pool_id.0, &who, share_amount)?;
```

Comment:
Seems to me `share_amount` could be below `MinPoolLiquidity` and thus be reaped directly. Does that violate any assumptions or is that fine? Could this be a problem for small liquidity providers?

Reply:
yes, potentially. as mentioned in the comment above - i will add check to make sure that share amount is always above that limit,.

Reply:
This should be solved in XYK. There is a reason (I hope it's documented) why it was done in such way and that is to prevent interlocking LPs from withdrawing. (I think it's checking whether you're leaving at least existential deposit there)

Reply:
Anyway I have another concern and that is when the trades go to extremes and the pool balance drops below ed is this going to reap the account?

Reply:
yes it is done in xyk like that.

actually it is done in this pallet too when liquidity is removed.

the only check is missing that newly minted shares are sufficient. Just adding that.

Reply:
Added.

Reply:
> Anyway I have another concern and that is when the trades go to extremes and the pool balance drops below ed is this going to reap the account?

how we prevent this in xyk ? 

we can do similar thing here too - i can;t recall now how it is done.

Reply:
I don't think we do actually

Reply:
> I don't think we do actually

created issue #490 to discuss/track/decide what needs to be done.